Cleanup PostgreSQL for state version 17.09

bachp opened the pull request:
###### Motivation for this change

- Update PostgreSQL to 9.6
- Change the superuser to be the more standard `postgres` user

Both these changes only apply if `system.stateVersion` is 17.09 or higher.

See: https://github.com/NixOS/nixpkgs/pull/18144

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bachp committed:
postgresql service: make 9.6 the default version for 17.09

bachp committed:
postgresql service: change default superuser for 17.09

Change the default superuser from `root` to `postgres` for state
version 17.09

Mic92 commented:
could be added to release notes.

zimbatm reviewed (commented):
Pretty good!

I think that all that's missing is a release note that explains the change and how to upgrade from 9.5 to 9.6 including changing the `createuser -s postgres`

bachp commented:
Can somebody point me to where the releasenote cam be updated?

LnL7 commented:
The [17.09 release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-1709.xml) are part of the nixos manual and you can build it with
`nix-build ./nixos/release.nix -A manual`

bachp commented:
I tried to add this the information to the release notes but I'm strugling to get the docbook syntax right. I first need to find some time to dig into this. So if somebody else is willing to write the entry please feel free to step up :wink:

basvandijk commented:
+1 on this change.

What about having the following change as well?

```nix
{
  dataDir = mkOption {
    default = 
      if versionAtLeast config.system.stateVersion "17.09" 
      then "/var/db/postgresql/${config.services.postgresql.package.psqlSchema}"
      else "/var/db/postgresql";
  };
}
```

Including the schema version (like "9.6") in the `dataDir` has the advantage that it's easier running multiple different versions of postgresql side by side. More importantly it makes running [pg_upgrade](https://www.postgresql.org/docs/current/static/pgupgrade.html) easier.

Ubuntu, Debian and Gentoo do this as well.

bachp commented:
@balajisivaraman I tought about this aswell. I f no objections are there I will add the schema version to the path.

bachp commented:
@balajisivaraman I tought about this aswell. I f no objections are there I will add the schema version to the path.

bachp committed:
postgresql service: change default data directory for 17.09

The new directory includes the schema version of the database.
This makes upgrades easier and is more consistent with other distros.

bachp commented:
I added the change proposed by @balajisivaraman

balajisivaraman commented:
@bachp I think you are mentioning me by mistake in this thread. Just letting you know so you can tag the correct person. 👍

bachp commented:
@balajisivaraman Sorry I did this on my mobile :wink: I actually wanted to mention @basvandijk.

edolstra commented:
Looks good to me.

zimbatm committed:
postgresql service: add release notes for changes

zimbatm commented:
Added some release notes to the branch. @bachp if you can rebase the branch I will merge it.

bachp commented:
I'm thinking if changing the data dir to `/var/db` makes sense. I would rather keep it in `/var/lib` as this is more common. See also discussion in #25931.

basvandijk commented:
@bachp I would also slightly prefer that the default is moved to `/var/lib/postgresql/$psqlSchema` since most applications store state under `/var/lib` and other distributions also store their PostgreSQL databases under `/var/lib`:

OS | $PGDATA
--------|--------------------------------------------
Ubuntu | `/var/lib/postgresql/$psqlSchema/$cluster`
Debian | `/var/lib/postgresql/$psqlSchema/$cluster`
Gentoo | `/var/lib/postgresql/$psqlSchema/data`
Arch | `/var/lib/postgres/data`
Fedora | `/var/lib/pgsql/data`

zimbatm commented:
I just noticed that the release note I added already mentions `/var/lib/postgres` as I assumed that we followed convention with other distros ^_^

bachp commented:
@zimbatm @basvandijk I rebase the PR and change the path to `/var/lib` so it is consistent with the changelog.

zimbatm commented on nixos/doc/manual/release-notes/rl-1709.xml:
```diff
@@ -78,6 +78,14 @@ rmdir /var/lib/ipfs/.ipfs
+      The <literal>postgres</literal> superuser name has changed from <literal>root</literal> to <literal>postgres</literal> to more closely follow what other Linux distributions are doing.
+    </para>
+    <para>
+      The <literal>postgres</literal> default <literal>dataDir</literal> has changed from <literal>/var/lib/postgres</literal> to <literal>/var/db/postgresql/$psqlSchema</literal> where $psqlSchema is 9.6 for example.
```
The rename described here is the wrong way around (my fault). It should be `from /var/db/postgres to /var/lib/postgres/$psqlSchema`

zimbatm committed:
fixup! postgresql service: add release notes for changes

zimbatm commented:
!!

bachp commented:
@zimbatm I will fix this together with a changelog entry for #25931
